src/hivpy/pregnancy.py

```diff
@@ -0,0 +1,268 @@
+                                           & (~pop.data[col.LOW_FERTILITY])
+                                           & (pop.data[col.AGE] >= 15)]
+        # group females by age groups
+        age_groups = np.digitize(pop.data.loc[female_population, col.AGE],
+                                 [15, 25, 35, 45])
+        pop.data.loc[female_population, col.AGE_GROUP] = age_groups
+        # outcomes
+        num_children = pop.transform_group([col.AGE_GROUP], self.calc_init_num_children_outcomes,
+                                           sub_pop=female_population)
+        # assign outcomes
+        pop.data.loc[female_population, col.NUM_CHILDREN] = num_children
+        # give everyone with a child a pregnancy date before the start of the simulation
+        pop.data.loc[pop.data[col.NUM_CHILDREN] > 0, col.LAST_PREGNANCY_DATE] = pop.date - timedelta(days=270)
+
+    def calc_init_num_children_outcomes(self, age_group, size):
+        """
+        Uses the probability distribution for a given age group to return
+        outcomes for each individual's initial number of children.
+        """
+        index = int(age_group)-1
+        outcomes = self.init_num_children_distributions[index].sample(size)
+        return outcomes
+
+    def update_pregnancy(self, pop):
+        """
+        Monitor pregnancies and model childbirth.
+        """
+        # TODO: this is needed for stp prob preg reduction but should
+        # probably be moved elsewhere so it's only initialised once
+        self.fold_tr_newp = pop.hiv_status.fold_tr_newp
+        # get sexually active female population to check for new pregnancies
+        active_female_population = pop.data[(pop.data[col.SEX] == SexType.Female)
+                                            & (pop.data[col.AGE] >= 15)
+                                            & (pop.data[col.AGE] < 55)
+                                            & (~pop.data[col.LOW_FERTILITY])
+                                            & (~pop.data[col.PREGNANT])
+                                            & (pop.data[col.NUM_CHILDREN] < self.max_children)
+                                            & ((pop.data[col.NUM_PARTNERS] > 0)
+                                                | pop.data[col.LONG_TERM_PARTNER])]
+
+        # continue if sexually active females are present this timestep
+        if len(active_female_population) > 0:
+
+            # at least 6 months must pass since last childbirth
+            # (15 months since last pregnancy start date)
+            previously_pregnant = active_female_population[active_female_population[col.LAST_PREGNANCY_DATE].notnull()]
+            pregnancy_ready = ~active_female_population[col.LOW_FERTILITY]
+            if len(previously_pregnant) > 0:
+                # mask
+                pregnancy_ready = pregnancy_ready & (previously_pregnant[col.LAST_PREGNANCY_DATE]
+                                                     + timedelta(days=450) <= pop.date)
```

Comment:
You're right, `~active_female_population[col.LOW_FERTILITY]` is redundant, I just needed it there to create an all True boolean series of the same length as `active_female_population`. Perhaps it would be less confusing to replace it with something like `[True] * len(active_female_population)`?

Also, I see what you're saying, but I don't think making the `pregnancy_ready` _and_ statement into an _or_ statement works because it would be the same as not checking whether a woman has been recently pregnant. The point is that the entire `active_female_population` should be capable of getting pregnant (which includes women who have never been pregnant and thus have a `None` last pregnancy date, otherwise my tests would be failing!), _unless_ they are currently in their "cooldown period" from being recently pregnant.

Reply:
Right, but the problem is what if we have a women who has `col.LAST_PREGNANCY_DATE` as `None`? Then we have `(previously_pregnant[col.LAST_PREGNANCY_DATE] + timedelta(days=450) <= pop.date)` is `False` and therefore `pregnancy_ready = True & False = False` don't we?